### PR TITLE
Cps/transonic mach number derivative

### PR DESCRIPTION
**Pull request by eloisabaez**
- Implementation in potential flow utilities of function to compute the mach number derivative WRT the velocity squared and corresponding tests
- Based on section A.2 of Fully Simulataneous Coupling of the Full Potential Equation and the Integral Boundary Layer Equations in Three Dimensions by Brian Nishida (1996)

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp [hunk 1]
```diff
@@ -416,6 +416,36 @@ bool CheckWakeCondition(const Element& rElement, const double& rTolerance, const
     return wake_condition_is_fulfilled;
 }
 
+template <int Dim, int NumNodes>
+double ComputeLocalMachSquaredDerivative(
+    const array_1d<double, Dim>& velocity,
+    const double& local_mach_number,
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]:
```suggestion
    const array_1d<double, Dim>& rVelocity,
    const double rLocalMachNumber,
```

File: applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp [hunk 2]
```diff
@@ -416,6 +416,36 @@ bool CheckWakeCondition(const Element& rElement, const double& rTolerance, const
...
+    // read free stream values
+    const double heat_capacity_ratio = rCurrentProcessInfo[HEAT_CAPACITY_RATIO];
+    const double free_stream_mach = rCurrentProcessInfo[FREE_STREAM_MACH];
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+
+    // make squares of values
+    double sq_local_mach_number = pow(local_mach_number, 2);
+    double sq_free_stream_mach = pow(free_stream_mach, 2);
+    double sq_local_velocity = inner_prod(velocity, velocity);
+    double sq_free_stream_velocity = inner_prod(free_stream_velocity, free_stream_velocity);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]:
```suggestion
    const double sq_local_mach_number = std::pow(local_mach_number, 2);
    const double sq_free_stream_mach = std::pow(free_stream_mach, 2);
    const double sq_local_velocity = inner_prod(velocity, velocity);
    const double sq_free_stream_velocity = inner_prod(free_stream_velocity, free_stream_velocity);
```

File: applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp [hunk 3]
```diff
@@ -416,6 +416,36 @@ bool CheckWakeCondition(const Element& rElement, const double& rTolerance, const
...
+
+    // square bracket term
+    double square_bracket_term = 1.0 + 0.5*(heat_capacity_ratio - 1)*
+            sq_free_stream_mach*(1 - (sq_local_velocity/sq_free_stream_velocity));
+
+    // calculate mach derivative
+    return sq_local_mach_number * ((1/sq_local_velocity) +
+            0.5*(heat_capacity_ratio - 1)*(1/sq_free_stream_velocity)*sq_free_stream_mach*(1/square_bracket_term));
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]:
```suggestion
    double square_bracket_term = 1.0 + 0.5*(heat_capacity_ratio - 1.0)*
            sq_free_stream_mach*(1.0 - (sq_local_velocity/sq_free_stream_velocity));

    // calculate mach derivative
    return sq_local_mach_number * ((1.0/sq_local_velocity) +
            0.5*(heat_capacity_ratio - 1.0)*(1.0/sq_free_stream_velocity)*sq_free_stream_mach*(1.0/square_bracket_term));
```

be carefuls with ints and doubles, mixing them can lead to unpleasant surprises

File: applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp (hunk 1 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]: I think @philbucher actually meant: const double& rLocalMachNumber

To follow the [style guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide#arguments): "Write arguments with a UpperCamelCase format. First input arguments, then output arguments. If the argument is passed as a reference, add 'r' at the beginning of the name. If the argument is a pointer, add 'p' at the beginning of the name."

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]: nope, I meant `const double rLocalMachNumber`

there is no reason to use a reference here, only confuses the reader

afaik both double and ref to double have a size of 64bits

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]: why then using r if it is not a reference?

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]: ah
sure, without the `r`

I was focused on the `&` :)

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp]: Ok!

> there is no reason to use a reference here, only confuses the reader
> 
> afaik both double and ref to double have a size of 64bits

that's actually a good point, I didn't know it

**Comment by eloisabaez** [comment]
thanks for the comments, I have incorporated them. But I think I will move what I have been working on into a single pull request as I think it makes more sense for this set of functions. Still figuring how to use github efficiently!

**Pull request closed without merging**